File: SessionInfoFinder.java
package com.asha.test;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

/**
 * This class finds the average session time and the most engaged user. (Goal 2 and Goal 4 in the task list).
 * this class uses the output of a mapreduce job (GetSessionTime: finds the latest visiting time for each user in a session, and generate a small output file), 
 * then runs on local computer.
 * Average session time is calculated by aggregating the time difference between the latest visiting time and session starting time.
 * Most engaged user is calculated by finding the user who has the longest session time.
 * @author asha
 *
 */
public class SessionInfoFinder {
	public static void main(String[] args) {

		String fileName = "/home/asha/hadoop-2.6.4/session_data/session_time";
		String line = null;
		int count = 0;
		long timeDiff = 0;
		String mostEngagedUser = "";

		SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.S'Z'");

		try {
			Date startTime = sdf.parse("2015-07-22T09:00:00.000Z");
			Date maxTime = sdf.parse("2015-07-22T09:00:00.000Z");
			FileReader fileReader = new FileReader(fileName);
			BufferedReader bufferedReader = new BufferedReader(fileReader);

			while ((line = bufferedReader.readLine()) != null) {
				count++;
				String[] results = line.split("\\s");
				Date currentTime = sdf.parse(results[1]);
				timeDiff += (currentTime.getTime() - startTime.getTime());
				if (currentTime.after(maxTime)) {
					mostEngagedUser = results[0];
					maxTime = currentTime;
				}
			}
			double diffSeconds = timeDiff / count / 1000.0;

			System.out.println("Avg session time: " + diffSeconds + "seconds");
			System.out.println("The most engaged user ip is:" + mostEngagedUser);
			bufferedReader.close();
		} catch (FileNotFoundException ex) {
			System.out.println("Unable to open file '" + fileName + "'");
		} catch (IOException ex) {
			System.out.println("Error reading file '" + fileName + "'");
		} catch (ParseException e) {
			System.out.println("Error parsing file '" + fileName + "'");
		}
	}

}

File: scala/src/main/scala/SessionInfoFinder.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf
import scala.io.Source
import java.io._
import java.text.SimpleDateFormat

/**
 * This class aggregates the unique visits for each URL in a session (Goal 3 in the task list)
 * Assume:
 *   - session time window is fixed, 15 minutes
 *   - this session time window starts at 2015-07-22T09:00:00.000Z
 *   - this session time windows ends at 2015-07-22T09:15:00.000Z
 * It runs as a mapreduce job on a hadoop cluster.
 * Mapper prepares the data as (url, ip)
 * Reducer removes the duplicated ips and count the unique visits.
 * @author asha
 *
 */
object UserInfo {

	def main(args: Array[String]) : Unit = {
		val filename = "sessionTime.out"
		val line = null
		var count = 0
		var timeDiff = 0L
		var mostEngagedUser = ""
		val sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.S'Z'")
		try 
		{
			val startTime = sdf.parse("2015-07-22T09:00:00.000Z")
			var maxTime = sdf.parse("2015-07-22T09:00:00.000Z")
			for (line <- Source.fromFile(filename).getLines()) 
			{
  				count += 1
				val results = line.split(",")
				val currentTime = sdf.parse(results(1).substring(0, results(1).size - 1))
				timeDiff += (currentTime.getTime() - startTime.getTime())
				if (currentTime.after(maxTime)) {
					mostEngagedUser = results(0).substring(1, results(0).size - 1)
					maxTime = currentTime
				}
			}
			val diffSeconds = timeDiff / count / 1000.0
			System.out.println("Avg session time: " + diffSeconds + "seconds")
			System.out.println("The most engaged user ip is:" + mostEngagedUser)
		} 
		catch 
		{
  			case ex: FileNotFoundException => println("Couldn't find that file.")
  			case ex: IOException => println("Had an IOException trying to read that file")
		}
	}
}
